### ISSUE TITLE
Rationalize request/validation meta in vuex modules

### ISSUE THREAD

**Issue by nicksellen** [yunity/karrot-frontend#691]
### Problem

We have lots of repeated action defintions like this:

```js
async fetchListByGroupId ({ commit }, groupId) {
    commit(types.REQUEST_LIST)
    try {
      commit(types.RECEIVE_LIST, { pickups: await pickups.listByGroupId(groupId), groupId })
    }
    catch (error) {
      commit(types.RECEIVE_LIST_ERROR, { error })
    }
  },
```

This is bad because:
- we might implement it slightly differently each time
- hard to improve across all files as lots to change
- makes it feel very heavy to add actions
- low signal/noise ratio (most of the code is boilerplate)

### Solution

There are two useful types of data here:
- meta data about the request, and validation
- longer lived _actual_ data

We want to separate these two concerns, so longer lived data would still be handled as it is, but we seperate all the boring request/validation meta data out into seperate nested vuex modules.

### Implementation

The parent store (e.g. `groups`) needs to have access to nested module getters to get the meta state, and can add it to the data object in an enrich method.

In a vuex module you would create a meta module, which does a few things:
- defines state and mutations for modifying meta about requests and validation (keyed by `metaId`)
- define a getter to get nice stuff by `metaId` (perhaps including a nice error mapper getter thing?)
- includes a function to define an action to be included in the main parent store thing (e.g. the `groups` module itself), which will call the appropriate mutations in the meta module

```js
const meta = createMetaModule()
export const modules = { meta } // I *think* that is how nested modules work?

export const getters = {
   enrich (state, getters) => pickup => {
    return { ...pickup, getters['meta/get'](pickup.id) }
  }
}

// .. blahblhablah

export const actions = {
  // "createRequestAction" returns a regular action definition
  leave: meta.createRequestAction({
    metaId: pickupId => pickupId,
    async request ({ dispatch }, pickupId) {
      await pickups.leave(pickupId)
      commit(types.RECEIVE_LEAVE, { pickupId, userId: rootGetters['auth/userId'] })
    },
  }),
}
```

`createRequestAction` takes a configuration object that wants a few things:
- `metaId` this is the key used to store the meta against, and what you have to use in the getter, can also be a fixed value too (e.g. in the case of meta about a `list` request)
- `request` a function that behaves exactly like a normal action (internally it is wrapped into something that handles error better)

The error handling needs to distinguish between request errors (400/500, etc... anything from axios), and programming errors in the bit of code _after_ the request.
- request errors should be handled (if they are validation errors then it processes that data and sets it nicely inside the meta module state, if they are other errors then we just re-throw them, something else will have shown some generic toast type thing)

**Comment by nicksellen** [comment]
Here is my earlier playing around with programmatically defining modules, might be useful to pick through (but implements a different concept):

```js
function defineModule ({ name, state: initialStateFn, run, after, mutations = {} }) {
  return {
    namespaced: true,
    state: initialStateFn(),
    actions: {
      async run (context, requestData) {
        let responseData
        let { commit } = context
        commit('request')
        try {
          responseData = await run(context, requestData)
          commit('receive', responseData)
        }
        catch (error) {
          commit('receiveError', { error })
          return
        }

        if (after) {
          await after(context, responseData)
        }
      },
      async clear (context, data) {
        let { commit } = context
        commit('clear')
      },
    },
    types: {
      request: `Request ${name}`,
      receive: `Receive ${name}`,
      receiveError: `Receive ${name} error`,
      clear: `Clear ${name}`,
    },
    mutations: {
      clear (state) {
        Object.assign(state, initialStateFn())
      },
      ...mutations,
    },
  }
}
```

**Comment by tiltec** [comment]
I worked on this a bit today, #693 came out of it.

I didn't like first implementation using `meta.createRequestAction`, as the commit happens in the wrong namespace (parent) and so I needed to do manual namespacing.

Using a parent action to dispatch actions to the child seems more natural to me and seems to fulfill all requirements.
https://github.com/yunity/karrot-frontend/blob/147c01a4e0f42fcf2917f73d797d60b10c74a2fe/src/store/modules/groups.js#L216-L225

The only remaining bit that I didn't like is the handling of composed request IDs:
https://github.com/yunity/karrot-frontend/blob/147c01a4e0f42fcf2917f73d797d60b10c74a2fe/src/store/modules/groups.js#L79-L83

It would be easy enough to just use `save` instead of `save-<id>` as a meta ID, so I would prefer that in this use case.
